stack/eslint-plugin-packlets/README.md

```diff
@@ -0,0 +1,166 @@
+
+- `src/logging/*.ts` - the logging system
+- `src/data-model/*.ts` - the data model
+- `src/reports/*.ts` - the report engine
+- `src/*.ts` - other arbitrary files such as startup code and the main application
+
+Using file folders is helpful, but it's not very strict. Files under `src/logging` can easily import files from `/src/reports`, creating a confusing circular import.  They can also import arbitrary application files.  Also, there is no clear distinction between which files are the "public API" for `src/logging` versus its private implementation details.
+
+All these problems can be solved by reorganizing the project into NPM packages (or [Rush projects](https://rushjs.io/)).  Something like this:
+
+- `@my-app/logging` - the logging system
+- `@my-app/data-model` - the data model
+- `@my-app/reports` - the report engine
+- `@my-app/application` - other arbitrary files such as startup code and the main application
+
+However, separating code in this ways has some downsides.  The projects need to build separately, which has some tooling costs (for example, "watch mode" now needs to consider multiple projects).  In a large monorepo, the library may attract other consumers, before the API has been fully worked out.
+
+Packlets provide a lightweight alternative that provides many of the same benefits of packages, but without the `package.json` file.  It's a great way to prototype your project organization before later graduating your packlets into proper NPM packages.
+
+## 5 rules for packlets
+
+With packlets, our folders would be reorganized as follows:
+
+- `src/packlets/logging/*.ts` - the logging system
+- `src/packlets/data-model/*.ts` - the data model
+- `src/packlets/reports/*.ts` - the report engine
+- `src/*.ts` - other arbitrary files such as startup code and the main application
+
+The [packlets-tutorial](https://github.com/microsoft/rushstack/tree/master/packlets/tutorials/packlets-tutorial) sample project shows this layout in more detail.
+
+The basic design can be summarized in 5 rules:
+
+1. A "packlet" is defined to be a folder path `./src/packlets/<packlet-name>/index.ts`. The **index.ts** file will have the exported APIs. The `<packlet-name>` name must consist of lower case words separated by hyphens, similar to an NPM package name.
+
+    Example file paths:
+    ```
+    src/packlets/controls
+    src/packlets/logger
+    src/packlets/my-long-name
+    ```
+
+    > **NOTE:** The `packlets` cannot be nested deeper in the tree. Like with NPM packages, `src/packlets` is a flat namespace.
+
+2. Files outside the packlet folder can only import the packlet root **index.ts**:
+
+    **src/app/App.ts**
+    ```ts
+    // Okay
+    import { MainReport } from '../packlets/reports';
+
+    // Error: The import statement does not use the packlet's entry point (@rushstack/packlets/mechanics)
```

Comment:
Please allow this. This is strictly superior to the directory import in terms of resolution accuracy and performance in TypeScript, Webpack, etc.

Reply:
If that's true (can you provide some details?), I feel like we should improve the tooling rather than asking developers to write their code in an awkward way.

Reply:
@dmichon-msft  I'm interested in this question, but I'm trying to unblock a couple teams who are waiting for these lint rules. So if you feel strongly about it, let's sort it out in a separate GitHub issue.

Reply:
Expecting people to always say `import { ... } from 'package-name/lib/index';` is awkward and will break SPFx development scenarios.

This isn't an optimization that we should expect developers to have to make because it produces objectively worse code for the sake of making the compiler slightly faster. If these filesystem accesses are a real concern, we should try to improve our caching and maybe explore filling a filesystem cache in a background thread.

Reply:
I'm fine with not making it the default. I'm not so happy about forbidding it. If we were actually writing code directly for browsers, we'd be expected to include the final step of doing `import { MainReport } from '../packlets/reports/index.js'`, but unfortunately TypeScript doesn't allow us to be that specific.

Reply:
we can continue this topic here: https://github.com/microsoft/rushstack/issues/2262